Lib/asyncio/base_events.py

```diff
@@ -867,6 +925,63 @@ def _check_sendfile_params(self, sock, file, offset, count):
 
         return transport, protocol
 
+    async def sendfile(self, transport, file, offset=0, count=None,
+                       *, fallback=True):
+        """Send a file through a transport.
+
+        Return amount of sent bytes.
+        """
+        if transport.is_closing():
+            raise RuntimeError("Transport is closing")
+        mode = getattr(transport, '_sendfile_compatible',
+                       constants._SendfileMode.UNSUPPORTED)
+        if mode is constants._SendfileMode.UNSUPPORTED:
+            raise RuntimeError(
+                f"sendfile is not supported for transport {transport!r}")
+        if mode is constants._SendfileMode.TRY_NATIVE:
+            try:
+                return await self._sendfile_native(transport, file,
+                                                   offset, count)
+            except events.SendfileNotAvailableError as exc:
+                if not fallback:
+                    raise
+        # the mode is FALLBACK or fallback is True
+        assert (mode is constants._SendfileMode.FALLBACK or mode is
+                constants._SendfileMode.TRY_NATIVE and fallback), mode
+        return await self._sendfile_fallback(transport, file,
+                                             offset, count)
+
+    async def _sendfile_native(self, transp, file, offset, count):
+        raise events.SendfileNotAvailableError(
+            "sendfile syscall is not supported")
+
+    async def _sendfile_fallback(self, transp, file, offset, count):
+        if offset:
+            file.seek(offset)
+        blocksize = min(count, 16384) if count else 16384
+        buf = bytearray(blocksize)
+        total_sent = 0
+        proto = _SendfileProtocol(transp)
+        try:
+            while True:
+                if count:
+                    blocksize = min(count - total_sent, blocksize)
+                    if blocksize <= 0:
+                        return total_sent
+                fut = proto._paused
+                if fut is not None:
```

Comment:
I think it would be more readable if `_SendfileProtocol` exposed a `drain` coro-method:

```python
await proto.drain()
...
transp.write(..)
```

Reply:
Where `_SendfileProtocol`'s drain would look like:

```python
async def drain():
    if self._can_write is None:
        return
    await self._can_write
    self._can_write = None
```

Reply:
No. 
The `_paused` attribute should be removed on writer resuming.
Doing it on draining is awkward and error prone.

Also I don't line `drain` name, it was always confusing.
We do access to attributes of other classes already in asyncio often enough.
A new *private* method for just replacing plain attribute access? Do we really need it?

Reply:
> A new private method for just replacing plain attribute access? Do we really need it?

Yes, because the code is super hard to read right now.  `drain` isn't really confusing in this context.

Since it's a private method and the code will certainly look nicer/tighter, then why not?